File: utilslib/src/main/java/com/ashlikun/utils/receiver/SmsReceiver.java
package com.ashlikun.utils.receiver;

import android.content.BroadcastReceiver;
import android.content.ContentValues;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.telephony.SmsManager;
import android.telephony.SmsMessage;

import com.ashlikun.utils.other.LogUtils;

import java.util.List;

/**
 * 作者　　: 李坤
 * 创建时间: 2017/8/7 10:05
 * 邮箱　　：496546144@qq.com
 * <p>
 * 功能介绍：短信接收器，可获取短信内容，发送者号码，短信中心号码等。
 * <uses-permission android:name="android.permission.RECEIVE_SMS"/>
 * action: android.provider.Telephony.SMS_RECEIVED
 */

public class SmsReceiver extends BroadcastReceiver {
    private static final String TAG = SmsReceiver.class.getSimpleName();
    private SmsListener smsListener;

    public SmsReceiver() {

    }

    @Override
    public void onReceive(Context context, Intent intent) {
        try {
            LogUtils.i("收到广播：" + intent.getAction());
            Bundle bundle = intent.getExtras();
            for (String key : bundle.keySet()) {
                LogUtils.i(key + " : " + bundle.get(key));
            }
            Object[] pdus = (Object[]) intent.getExtras().get("pdus");
            String fromAddress = null;
            String serviceCenterAddress = null;
            if (pdus != null) {
                String msgBody = "";
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.DONUT) {
                    for (Object obj : pdus) {
                        SmsMessage sms = SmsMessage.createFromPdu((byte[]) obj);
                        msgBody += sms.getMessageBody();
                        fromAddress = sms.getOriginatingAddress();
                        serviceCenterAddress = sms.getServiceCenterAddress();

                        if (smsListener != null) {
                            smsListener.onMessage(sms);
                        }
                    }
                }
                if (smsListener != null) {
                    smsListener.onMessage(msgBody, fromAddress, serviceCenterAddress);
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void registerSmsReceiver(Context context, SmsListener smsListener) {
        try {
            this.smsListener = smsListener;
            IntentFilter filter = new IntentFilter();
            filter.addAction("android.provider.Telephony.SMS_RECEIVED");
            filter.setPriority(Integer.MAX_VALUE);
            context.registerReceiver(this, filter);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void unRegisterSmsReceiver(Context context) {
        try {
            context.unregisterReceiver(this);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static abstract class SmsListener {
        public abstract void onMessage(String msg, String fromAddress, String serviceCenterAddress);

        public void onMessage(SmsMessage msg) {
        }
    }

    /**
     * Call requires API level 4
     * <uses-permission android:name="android.permission.SEND_SMS"/>
     *
     * @param phone
     * @param msg
     */
    public static void sendMsgToPhone(String phone, String msg) {
        LogUtils.i("发送手机：" + phone + " ,内容： " + msg);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.DONUT) {
            SmsManager manager = SmsManager.getDefault();
            List<String> texts = manager.divideMessage(msg);
            for (String txt : texts) {
                manager.sendTextMessage(phone, null, txt, null, null);
            }
        } else {
            LogUtils.i("发送失败，系统版本低于DONUT，" + phone + " ,内容： " + msg);
        }

    }

    /**
     * Call requires API level 4
     * <uses-permission android:name="android.permission.WRITE_SMS"/>
     *
     * @param context
     * @param phone
     * @param msg
     */
    public static void saveMsgToSystem(Context context, String phone, String msg) {
        ContentValues values = new ContentValues();
        values.put("date", System.currentTimeMillis());
        //阅读状态 
        values.put("read", 0);
        //1为收 2为发  
        values.put("type", 2);
        values.put("address", phone);
        values.put("body", msg);
        context.getContentResolver().insert(Uri.parse("content://sms/inbox"), values);
    }

}

File: utilslib/src/main/java/com/ashlikun/utils/receiver/SmsReceiver.kt
package com.ashlikun.utils.receiver

import android.content.*
import android.net.Uri
import android.os.Build
import android.telephony.SmsManager
import android.telephony.SmsMessage
import com.ashlikun.utils.other.LogUtils.i

/**
 * @author　　: 李坤
 * 创建时间: 2021/12/13 23:53
 * 邮箱　　：496546144@qq.com
 *
 * 功能介绍：短信接收器，可获取短信内容，发送者号码，短信中心号码等。
 * <uses-permission android:name="android.permission.RECEIVE_SMS"></uses-permission>
 * action: android.provider.Telephony.SMS_RECEIVED
 */
typealias SmsListener = (msg: SmsMessage) -> Unit
typealias SmsListenerLast = (msg: String, fromAddress: String, serviceCenterAddress: String) -> Unit

class SmsReceiver : BroadcastReceiver() {
    var smsListener: SmsListener? = null
    var smsListenerLast: SmsListenerLast? = null
    override fun onReceive(context: Context, intent: Intent) {
        try {
            i("收到广播：" + intent.action)
            val bundle = intent.extras
            for (key in bundle!!.keySet()) {
                i(key + " : " + bundle[key])
            }
            val pdus = intent.extras!!["pdus"] as Array<Any>?
            var fromAddress = ""
            var serviceCenterAddress = ""
            if (pdus != null) {
                var msgBody = ""
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.DONUT) {
                    for (obj in pdus) {
                        val sms = SmsMessage.createFromPdu(obj as ByteArray)
                        msgBody += sms.messageBody ?: ""
                        fromAddress = sms.originatingAddress ?: ""
                        serviceCenterAddress = sms.serviceCenterAddress
                        smsListener?.invoke(sms)
                    }
                }
                smsListenerLast?.invoke(msgBody, fromAddress, serviceCenterAddress)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun registerSmsReceiver(context: Context, smsListener: SmsListener?) {
        try {
            this.smsListener = smsListener
            val filter = IntentFilter()
            filter.addAction("android.provider.Telephony.SMS_RECEIVED")
            filter.priority = Int.MAX_VALUE
            context.registerReceiver(this, filter)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun unRegisterSmsReceiver(context: Context) {
        try {
            context.unregisterReceiver(this)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }


    companion object {
        fun sendMsgToPhone(phone: String, msg: String) {
            i("发送手机：$phone ,内容： $msg")
            val manager = SmsManager.getDefault()
            val texts: List<String> = manager.divideMessage(msg)
            for (txt in texts) {
                manager.sendTextMessage(phone, null, txt, null, null)
            }

        }

        fun saveMsgToSystem(context: Context, phone: String?, msg: String?) {
            val values = ContentValues()
            values.put("date", System.currentTimeMillis())
            //阅读状态 
            values.put("read", 0)
            //1为收 2为发  
            values.put("type", 2)
            values.put("address", phone)
            values.put("body", msg)
            context.contentResolver.insert(Uri.parse("content://sms/inbox"), values)
        }
    }
}